### PR TITLE
executor: TiFlash supports stale read

### LINKED ISSUE #40047: executor: TiFlash supports stale read
## Enhancement
https://github.com/pingcap/tiflash/issues/4483

### PR DESCRIPTION
**Pull request by hehechen**
Signed-off-by: hehechen <chentongli@pingcap.com>

### What problem does this PR solve?

Issue Number: close #40047

Problem Summary:

### What is changed and how it works?
1. remove TiFlash stale read check.
2. use `query_ts`, `local_query_id`, `server_id` as the MPP query id.
3. [MinSafeTS in client-go has supported TiFlash.](https://github.com/tikv/client-go/pull/642)
4. [TiFlash code](https://github.com/pingcap/tiflash/pull/6459) has supported stale read.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
run CH-benCHmark with stale read
image
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
TiFlash supports stale read.
```

### CODE REVIEW

File: store/copr/mpp.go [hunk 1]
```diff
@@ -334,7 +336,7 @@ func (m *mppIterator) cancelMppTasks() {
 	m.mu.Lock()
 	defer m.mu.Unlock()
 	killReq := &mpp.CancelTaskRequest{
-		Meta: &mpp.TaskMeta{StartTs: m.startTs},
+		Meta: &mpp.TaskMeta{StartTs: m.startTs, QueryTs: m.rootDestinationTask.QueryTs, LocalQueryId: m.rootDestinationTask.LocalQueryID, ServerId: m.rootDestinationTask.ServerID},
```
Comment (SeaRise) [inline store/copr/mpp.go:339]: I prefer to wrap a struct like `MPPQueryId` instead of using `rootDestinationTask`.

File: executor/builder.go [hunk 2]
```diff
@@ -130,6 +132,8 @@ func newExecutorBuilder(ctx sessionctx.Context, is infoschema.InfoSchema, ti *Te
 		isStaleness:      staleread.IsStmtStaleness(ctx),
 		txnScope:         txnManager.GetTxnScope(),
 		readReplicaScope: txnManager.GetReadReplicaScope(),
+		queryTS:          uint64(time.Now().UnixNano()),
+		localQueryID:     plannercore.AllocMPPQueryID(),
```
Comment (SeaRise) [inline executor/builder.go:136]: Using `AllocMPPQueryID` in a non-mpp executor seems strange.
Can there be a more generic name?

Comment (SeaRise) [reply executor/builder.go:136]: I'm not sure it's a good idea to hold `queryTS` and `queryTS` in ExecutorBuilder

File: DEPS.bzl [hunk 3]
```diff
@@ -3523,12 +3523,20 @@ def go_deps():
         sum = "h1:mbAskLJ0oJfDRtkanvQPiooDH8HvJ2FBh+iKT/OmiQQ=",
         version = "v0.0.0-20181126055449-889f96f722a2",
     )
+    go_repository(
+        name = "com_github_tiancaiamao_gp",
```
[the remark this answers is not in the document]
Comment (hehechen) [reply DEPS.bzl:3527]: > Is this expected?Ref a private repo?

Yes, it's generated from go.mod.

File: executor/builder.go (hunk 2 above)
Comment (windtalker) [reply executor/builder.go:136]: Even for non-mpp query, `plannercore.AllocMPPQueryID()` will be called?

File: kv/mpp.go [hunk 4]
```diff
@@ -30,19 +30,25 @@ type MPPTaskMeta interface {
...
-	ID      int64       // mppTaskID
-	StartTs uint64
-	TableID int64 // physical table id
+	Meta         MPPTaskMeta // on which store this task will execute
+	ID           int64       // mppTaskID
+	StartTs      uint64
+	QueryTs      uint64 // timestamp of query execution, used for TiFlash minTSO schedule
+	LocalQueryID uint64 // unique mpp query id in local tidb memory.
+	ServerID     uint64
+	TableID      int64 // physical table id
```
Comment (windtalker) [inline kv/mpp.go:39]: `gather_id` is not considered?

File: planner/core/fragment.go [hunk 5]
```diff
@@ -79,29 +81,48 @@ type tasksAndFrags struct {
...
+		startTS:      startTs,
+		queryTS:      queryTs,
+		localQueryID: localQueryID,
+		is:           is,
+		cache:        make(map[int]tasksAndFrags),
 	}
 	return g.generateMPPTasks(sender)
 }
 
+var mppTaskID int64 = 1
```
Comment (windtalker) [inline planner/core/fragment.go:106]: `mppTaskID` is a global var and keep increasing?

Comment (hehechen) [reply planner/core/fragment.go:106]: Yes, just keep increasing is enough.

File: kv/mpp.go (hunk 4 above)
Comment (hehechen) [reply kv/mpp.go:39]: > `gather_id` is not considered?

`gather_id` is not relevant with stale_read, so not considered in this PR.

File: executor/mpp_gather.go [hunk 6]
```diff
@@ -38,13 +39,28 @@ func useMPPExecution(ctx sessionctx.Context, tr *plannercore.PhysicalTableReader
...
 // MPPGather dispatch MPP tasks and read data from root tasks.
 type MPPGather struct {
 	// following fields are construct needed
 	baseExecutor
 	is           infoschema.InfoSchema
 	originalPlan plannercore.PhysicalPlan
 	startTS      uint64
+	mppQueryID   kv.MPPQueryID
+	queryTS      uint64
+	localQueryID uint64
```
Comment (SeaRise) [inline executor/mpp_gather.go:63]: Seems useless.

**Review by SeaRise — APPROVED** [review]
others LGTM

**Review by XuHuaiyu — APPROVED** [review]

File: kv/mpp.go [hunk 7]
```diff
@@ -27,21 +27,32 @@ type MPPTaskMeta interface {
...
 // MPPTask means the minimum execution unit of a mpp computation job.
 type MPPTask struct {
-	Meta    MPPTaskMeta // on which store this task will execute
-	ID      int64       // mppTaskID
-	StartTs uint64
-	TableID int64 // physical table id
+	Meta       MPPTaskMeta // on which store this task will execute
+	ID         int64       // mppTaskID
+	StartTs    uint64
+	MppQueryID MPPQueryID
```
Comment (winoros) [inline kv/mpp.go:42]:
```suggestion
	MppQueryID MPPQueryID // Used for stale read.
```

File: kv/mpp.go [hunk 8]
```diff
@@ -70,10 +81,11 @@ type MPPDispatchRequest struct {
 	IsRoot  bool        // root task returns data to tidb directly.
 	Timeout uint64      // If task is assigned but doesn't receive a connect request during timeout, the task should be destroyed.
 	// SchemaVer is for any schema-ful storage (like tiflash) to validate schema correctness if necessary.
-	SchemaVar int64
-	StartTs   uint64
-	ID        int64 // identify a single task
-	State     MppTaskStates
+	SchemaVar  int64
+	StartTs    uint64
+	MppQueryID MPPQueryID
```
Comment (winoros) [inline kv/mpp.go:86]:
```suggestion
	MppQueryID MPPQueryID // Used for stale read.
```

File: executor/mpp_gather.go [hunk 9]
```diff
@@ -38,13 +39,26 @@ func useMPPExecution(ctx sessionctx.Context, tr *plannercore.PhysicalTableReader
...
+}
+
 // MPPGather dispatch MPP tasks and read data from root tasks.
 type MPPGather struct {
 	// following fields are construct needed
 	baseExecutor
 	is           infoschema.InfoSchema
 	originalPlan plannercore.PhysicalPlan
 	startTS      uint64
+	mppQueryID   kv.MPPQueryID
```
Comment (winoros) [inline executor/mpp_gather.go:61]:
```suggestion
	mppQueryID   kv.MPPQueryID // Used for stale read.
```

**Review by winoros — APPROVED** [review]

**Pull request merged**